src/Raven.Server/Documents/PeriodicBackup/Restore/ShardedRestoreOrchestrationTask.cs

```diff
@@ -0,0 +1,177 @@
+namespace Raven.Server.Documents.PeriodicBackup.Restore
+{
+    public class ShardedRestoreOrchestrationTask : AbstractRestoreBackupTask
+    {
+        private IOperationResult _result;
+
+        public ShardedRestoreOrchestrationTask(ServerStore serverStore,
+            RestoreBackupConfigurationBase configuration,
+            OperationCancelToken operationCancelToken) : base(serverStore, configuration, restoreSource: null, filesToRestore: null, operationCancelToken)
+        {
+        }
+
+        protected override async Task Initialize()
+        {
+            await base.Initialize();
+            CreateRestoreSettings();
+        }
+
+        protected override async Task OnBeforeRestore()
+        {
+            ModifyDatabaseRecordSettings();
+            InitializeShardingConfiguration();
+
+            var databaseRecord = RestoreSettings.DatabaseRecord;
+
+            // we are currently restoring, shouldn't try to access it
+            databaseRecord.DatabaseState = DatabaseStateStatus.RestoreInProgress;
+            await SaveDatabaseRecordAsync(DatabaseName, databaseRecord, RestoreSettings.DatabaseValues, Result, Progress);
+        }
+
+        protected override async Task Restore()
+        {
+            var dbSearchResult = ServerStore.DatabasesLandlord.TryGetOrCreateDatabase(DatabaseName);
+            var shardedDbContext = dbSearchResult.DatabaseContext;
+
+            _result = await RestoreOnAllShards(Progress, RestoreConfiguration, shardedDbContext);
+        }
+
+        protected override void OnAfterRestore()
+        {
+            DisableOngoingTasksIfNeeded(RestoreSettings.DatabaseRecord);
+
+            Progress.Invoke(Result.Progress);
+
+            // after the db for restore is done, we can safely set the db state to normal and write the DatabaseRecord
+            using (ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext serverContext))
+            using (serverContext.OpenReadTransaction())
+            {
+                var databaseRecord = ServerStore.Cluster.ReadDatabase(serverContext, DatabaseName);
+                databaseRecord.DatabaseState = DatabaseStateStatus.Normal;
+
+                RestoreSettings.DatabaseRecord = databaseRecord;
+            }
+        }
+
+        protected override IOperationResult OperationResult() => _result;
+
+        private void InitializeShardingConfiguration()
+        {
+            var databaseRecord = RestoreSettings.DatabaseRecord;
+
+            databaseRecord.Sharding = new ShardingConfiguration
+            {
+                Shards = new DatabaseTopology[RestoreConfiguration.ShardRestoreSettings.Length],
+                Orchestrator = new OrchestratorConfiguration
+                {
+                    Topology = new OrchestratorTopology
+                    {
+                        Members = new List<string>()
+                    }
+                }
+            };
+
+            var nodes = new HashSet<string>();
```

Comment:
do we want to have here OrdinalIgnoreCase comparer?

Reply:
done